Batches explore

0xaaiden opened the pull request:
# Summary

Added recent batches tab to the Homepage table

Screenshot:
![image](https://github.com/cowprotocol/explorer/assets/30964272/71bbeb83-2b1e-472b-bd7f-8b1f2be65a91)


# To Test

* Open homepage, click on Recent Batches Tab.
* Check it display recent batch orders
* Click on Tx. Hash should redirect to order page
* Hovering over tokens should take to etherscan token page
* Embedded Menu should have 3 items (Etherscan, Cow Explorer, and copy Tx Id)

0xaaiden committed:
adding recent batches tab to the table

0xaaiden committed:
fixing bugs, adding menu items

0xaaiden committed:
remove yarn.lock extra

0xaaiden committed:
Merge branch 'cowprotocol:develop' into batches-explore

fleupold commented:
This looks awesome! 

What is the logic for calculating the batch value? I noticed for a lot of them it says 0 even if e.g. ETH is involved. Can we infer the batch value if we have the price for at least one side of each trade? Also, in case we can't can we use something like "prices unknown" or similar rather than $0 ?

elena-zh commented:
Hey @0xaaiden , great job!
 I have done some tests and here are my notes: 
1. It would be great to start the 2nd word from a lower-case letter (as everywhere in the app)
![low case](https://github.com/cowprotocol/explorer/assets/70885163/c04ad071-c2ef-44c9-a97e-d376ef225ba0)
![small](https://github.com/cowprotocol/explorer/assets/70885163/d57ebb68-5a64-476f-9a7c-5d44321a1b40)

2. It would be great to crop a TX hash in the middle 
![cropp in the middle](https://github.com/cowprotocol/explorer/assets/70885163/75ec4ead-2810-4db6-903f-a5aa9dba1425)
3. It would be great to show a public name tag for a token (and give a bit more context for a token instead of showing 1 letter) 
![fetch token name](https://github.com/cowprotocol/explorer/assets/70885163/c14c229e-b49a-4971-80c3-3f012e15af95)
![tracker](https://github.com/cowprotocol/explorer/assets/70885163/90d4d749-a939-458c-94d0-8c215aff3234)
4. It would be great to show a downward arrow near the Date column to indicate the sorting
5. In a mobile view it says that the table is sorted by volume. Need to change to Date
![mobile sorting](https://github.com/cowprotocol/explorer/assets/70885163/2e857238-bd4d-4dd4-a536-f591ed7386a9)
6. It would be auto-update the table (once per 30 sec?) when it is opened to show recent batces (not using manual page refresh)
![autoupdate](https://github.com/cowprotocol/explorer/assets/70885163/2cf4887d-852b-4a95-b16b-51c6546d7069)

7. Menu fonts looks like broken 
![fonts](https://github.com/cowprotocol/explorer/assets/70885163/88dca033-72b6-4c43-a16c-27be461a6d15)
8. It would be great to add a copy icon closer to the TX hash. See how it is implemented for order/TX details pages: 
![copy](https://github.com/cowprotocol/explorer/assets/70885163/283a87ed-7b96-4cc3-88da-8ed88e72f150)
9. 'View on Etherscan' in the menu sounds incorrect for Gnosis chain. It should be 'View on Gnosisscan'
10. Search is not working in the table
![search is not working - disable](https://github.com/cowprotocol/explorer/assets/70885163/ad7e85b4-8654-4764-9d15-316915fa7916)
11. It would be great to add an anchor to the link and do not navigate a user to the Top tokens tab when refresh the page
![image](https://github.com/cowprotocol/explorer/assets/70885163/a9d01033-ce1c-4a89-8fb6-26cd80ee0ac9)
12. Token icons with amount are broken on order details/user details/TX details pages in this PR: 
![TX details](https://github.com/cowprotocol/explorer/assets/70885163/ab4d52d9-7343-4a06-9598-1ace811011c9)
13. No USD estimation for Gnosis chain for a batch value
![image](https://github.com/cowprotocol/explorer/assets/70885163/351cdcc9-f9b6-4d76-ade0-94f1b231f502)
14. I think it would be great to show a full date on hover/click on the Date column. Something like 
![date like](https://github.com/cowprotocol/explorer/assets/70885163/43c3b7c4-50b4-4a5a-9ada-84607c9faa34)
15. I'd add a bit more data in a mobile view. E.g. to show a token name close to a  token symbol. Also. menu looks weird as the last column there. 
![menu, add symbols](https://github.com/cowprotocol/explorer/assets/70885163/15882865-c23d-4739-96c2-fe699449599f)

0xaaiden committed:
fix changes, fix search Bar, fix display, responsiveness

0xaaiden committed:
Merge branch 'batches-explore' of https://github.com/0xaaiden/explorer-1 into batches-explore

0xaaiden committed:
fix copy button on recent batches table

0xaaiden committed:
fix menu font

0xaaiden commented:
Thanks @fleupold and @elena-zh for the detailed and extensive feedback (keep it coming!)
I've made all the styling changes mentioned above and fixed the search bar (filter applied on Tx. hash and solver ids).
@fleupold - for computing the batch value - it's now showing "Not Available" if the USD prices are not available for both the sell token or buy token. 

@elena-zh - I've removed the menu in mobile display as it was cluttering and didn't want to surcharge the screen. I've also added hot refetching every 30 seconds to update the table with the latest batches

I've attached a video below summarizing the changes.

https://github.com/cowprotocol/explorer/assets/30964272/fe72abe6-2cf5-44da-89a3-0b667cafeff9

anxolin commented on .env.example:
```diff
@@ -1,13 +1,13 @@
+AUTOCONNECT=true	
 
-# Public IPFS gateway
-REACT_APP_IPFS_READ_URI=https://cloudflare-ipfs.com/ipfs
```
why the whitespaces?

0xaaiden replied:
Not sure why it happened!$, fixed!

anxolin commented on src/apps/explorer/components/TokensTableWidget/BatchesTableWithData.tsx:
```diff
@@ -0,0 +1,41 @@
+    return (): void => {
+      clearTimeout(timeOutId)
+    }
+  }, [batches, batches?.length])
```
why do we need `batches?.length]` isn't batches enough?

anxolin commented on src/apps/explorer/components/TokensTableWidget/BatchesTableWithData.tsx:
```diff
@@ -0,0 +1,41 @@
+
+    const timeOutId: NodeJS.Timeout = setTimeout(() => {
+      setIsFirstLoading(false)
+    }, timeOutMs)
```
these logics seems overcomplicated, plus is strange that you are using this for the loading, but is only based on a timer... 

Wouldn't be cleaner design if the context would have a `loading: boolean` flag ? so responsibility is not in this component (the loading logic is not here anyways, so it doesn't feel is this component responsibility)

0xaaiden replied:
That's the same logic used to load the Tokens too

anxolin commented on src/apps/explorer/components/TokensTableWidget/BatchesTableWithData.tsx:
```diff
@@ -0,0 +1,41 @@
+    }
+  }, [batches, batches?.length])
+
+  return isFirstRender || isFirstLoading ? (
```
do you handle the errors too?

0xaaiden replied:
It's handled in the batchesTable component,
```ts
if (!items || items.length === 0) {
      tableContent = (
        <tr className="row-empty">
          <td className="row-td-empty">
            <EmptyItemWrapper>
              No results found. <br /> Please try another search.
            </EmptyItemWrapper>
          </td>
        </tr>
```
 If there is not data returned, it shows "No results found."

anxolin commented on src/apps/explorer/components/TokensTableWidget/index.tsx:
```diff
@@ -17,7 +20,12 @@ import CowLoading from 'components/common/CowLoading'
+const TABS = {
+  TOKENS: 1,
+  BATCHES: 2,
+}
```
can we use a enum?

anxolin commented on src/apps/explorer/components/TokensTableWidget/index.tsx:
```diff
@@ -71,39 +79,52 @@ const ExtraComponentNode: React.ReactNode = (
 )
+interface Props {
+  networkId: BlockchainNetwork
+}
```
nit: can you give a not so generic name for these props? 
you use it in `TokensTableWidget` so what about `TokensTableWidgetProps` and export it so other components can use it

anxolin replied:
also, see next lines. Looks like a duplicate

anxolin commented on src/apps/explorer/components/TokensTableWidget/index.tsx:
```diff
@@ -121,6 +142,14 @@ export const TokensTableWidget: React.FC<Props> = () => {
+    const handleHashChange = (): void => {
+      setSelectedTab(location.hash === '#batches' ? TABS.BATCHES : TABS.TOKENS)
+    }
+    window.addEventListener('hashchange', handleHashChange)
```
mmmm, is this the reactRouter way of doing it?

anxolin commented on src/apps/explorer/components/TokensTableWidget/index.tsx:
```diff
@@ -121,6 +142,14 @@ export const TokensTableWidget: React.FC<Props> = () => {
 
+  useEffect(() => {
+    const handleHashChange = (): void => {
+      setSelectedTab(location.hash === '#batches' ? TABS.BATCHES : TABS.TOKENS)
```
if you have an enum for this, you can use it also for checking the condition, so no need to add the string `'#batches'`

anxolin commented on src/apps/explorer/components/TokensTableWidget/index.tsx:
```diff
@@ -143,12 +176,19 @@ export const TokensTableWidget: React.FC<Props> = () => {
   }
-
+  const ExtraComponentNodeSearchBar: React.ReactNode = (
+    <div style={{ marginRight: '1rem' }}>
```
pls, don't hardcode CSS

anxolin commented on src/apps/explorer/components/TokensTableWidget/index.tsx:
```diff
@@ -143,12 +176,19 @@ export const TokensTableWidget: React.FC<Props> = () => {
       <TokensTableContext.Provider
         value={{
           data: filterData(),
+          batchesData: filterBatches(),
```
do we want to execute these function in each re-render?

anxolin commented on src/apps/explorer/components/TokensTableWidget/index.tsx:
```diff
@@ -160,7 +200,17 @@ export const TokensTableWidget: React.FC<Props> = () => {
+          tabItems={tabItems()}
+          updateSelectedTab={(tabId: number): void => {
+            setSelectedTab(tabId)
+            history.push(tabId === TABS.BATCHES ? '#batches' : '#tokens')
```
here you should use the enum... also can you do it in a more reactRouter way?

anxolin commented on src/hooks/useGetBatches.ts:
```diff
@@ -0,0 +1,244 @@
+  })
+  // Remove duplicates
+  return tokens.filter((token, index) => tokens.findIndex((t) => t.address === token.address) === index)
+}
```
i didn't review fully this file. Needs to be cleaned up. Too big and hard to review.

anxolin replied:
Also, i believe thegraph types should be auto-generated. 

Actually, this should already be part of the SDK (the autogeneration of types)

anxolin reviewed (commented):
Thanks for your contribution!!

anxolin commented:
One more thing, I actually wouldn't mind to make either the last batches the entry point, or to show both somehow (change the design slightly)

Easier is the first option. Just pinging @fairlighteth who might have an oppinion

elena-zh commented:
Hey @0xaaiden , great job!
Still I have some nitpicks: 
1. Mobile view, 2 tokens in the batch: it would be gear to add a space between tokens
![space](https://github.com/cowprotocol/explorer/assets/70885163/0f65a4b0-eee8-4dd4-92b6-bed6e57f7adf)
2. CoW in the Explorer should be with 'W'
![cow](https://github.com/cowprotocol/explorer/assets/70885163/29502bac-be3e-49da-b1ef-698bda18b000)
3. Goerli: no 'View on Etherscan' link that navigates to 'goerli.etherscan.tx_hash'
![goerli - TX on Etherscan or hide](https://github.com/cowprotocol/explorer/assets/70885163/650ec67d-6623-46ba-b2f4-af88cb2ea0b7)
4. When change a network, the app stays on the Recent batches tab, but the anchor is removed from the link. I think, we should keep it :) 
![not remove anchor](https://github.com/cowprotocol/explorer/assets/70885163/fce55d13-7227-4f12-816e-fa1b7c803663)
5. It would be great to indicate in the search field available search options 9n the placeholder. Currently, it shows a value for Tokens table
6. I have not been able to search by a solver address (however, IDK its full hash, just tried to search by this, and was not successful
![image](https://github.com/cowprotocol/explorer/assets/70885163/3ca53eaf-c778-4415-bcf8-5f2b958e9475)
7. I'd hide Copy action from the menu as currently it does nothing but navigating a user to the top on the Home page. Besides, not we have an icon neat a TX hash. 

Thanks!

0xaaiden committed:
Delete .env.example

0xaaiden committed:
Revert "Delete .env.example"

This reverts commit f67e5b619ba203b416342346ae440787eb565c35.

0xaaiden committed:
revert env example

shoom3301 commented:
Hi!
We don't see any activity in the PR for a long time.
Unfortunately, we have to close the PR since we are moving to another repo: https://github.com/cowprotocol/cowswap/pull/3579.
We will be happy if you continue the work here.
Thank you!
